pkg/agent/nodeportlocal/portcache/port_table_windows.go

```diff
@@ -0,0 +1,252 @@
+//go:build windows
+// +build windows
+
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package portcache
+
+import (
+	"fmt"
+
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/nodeportlocal/rules"
+)
+
+// podIPPortFormat formats the ip, port to string ip:port.
+func podIPPortProtoFormat(ip string, port int, protocol string) string {
+	return fmt.Sprintf("%s:%d:%s", ip, port, protocol)
+}
+
+func (pt *PortTable) getEntryByPodIPPortProto(ip string, port int, protocol string) *NodePortData {
+	return pt.PodEndpointTable[podIPPortProtoFormat(ip, port, protocol)]
+}
+
+func (pt *PortTable) GetEntry(ip string, port int, protocol string) *NodePortData {
+	pt.tableLock.RLock()
+	defer pt.tableLock.RUnlock()
+	// Return pointer to copy of data from the PodEndpointTable.
+	if data := pt.getEntryByPodIPPortProto(ip, port, protocol); data != nil {
+		dataCopy := *data
+		return &dataCopy
+	}
+	return nil
+}
+
+func addRuleForPort(podPortRules rules.PodPortRules, port int, podIP string, podPort int, protocol string) ([]ProtocolSocketData, error) {
+	// port needs to be available for all supported protocols: all available protocols
+	// should be returned if any NetNatStaticMapping rule can be inserted to an unused port.
+	protocols := make([]ProtocolSocketData, 0, len(supportedProtocols))
+	err := podPortRules.AddRule(port, podIP, podPort, protocol)
+	if err != nil {
+		klog.ErrorS(err, "Local port cannot be opened", "port", port, "protocol", protocol)
+		return nil, err
+	}
+	protocols = append(protocols, ProtocolSocketData{
+		Protocol: protocol,
+		State:    stateInUse,
+		socket:   nil,
+	})
+	return protocols, nil
+}
+
+func updateCloseState(protocols []ProtocolSocketData) error {
+	for idx := range protocols {
+		protocolSocketData := &protocols[idx]
+		if protocolSocketData.State != stateOpen {
+			continue
+		}
+		protocolSocketData.State = stateClosed
+	}
+	return nil
+}
+
+func (pt *PortTable) addRuleforFreePort(podIP string, podPort int, protocol string) (int, []ProtocolSocketData, error) {
+	klog.V(2).InfoS("Looking for free Node port on Windows", "podIP", podIP, "podPort", podPort, "protocol", protocol)
+	numPorts := pt.EndPort - pt.StartPort + 1
+	for i := 0; i < numPorts; i++ {
+		port := pt.PortSearchStart + i
+		if port > pt.EndPort {
+			// handle wrap around
+			port = port - numPorts
+		}
+		if _, ok := pt.NodePortTable[port]; ok {
```

Comment:
You may skip a Node port if it is already taken by a different protocol?

Reply:
Updated.